File: examples/rust/examples/client_collection_management.rs
//! Collection management using ekodb_client library
//!
//! This example demonstrates collection operations using the ekodb_client library.
//!
//! Compare with collection_management.rs to see the difference!

use ekodb_client::{Client, Record};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Create ekoDB client
    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("✓ Client created");

    let collection = "client_collection_management_rust";

    // Example 1: Create collection (via insert)
    println!("\n=== Create Collection (via insert) ===");
    let mut record = Record::new();
    record.insert("name", "Demo Record");
    record.insert("value", 42);

    let inserted = client.insert(collection, record).await?;
    if let Some(ekodb_client::FieldType::String(id)) = inserted.get("id") {
        println!("Collection created with first record: {:?}", id);
    }

    // Example 2: List all collections
    println!("\n=== List Collections ===");
    let collections = client.list_collections().await?;
    println!("Total collections: {}", collections.len());
    println!(
        "Sample collections: {:?}",
        collections.iter().take(5).collect::<Vec<_>>()
    );

    // Example 3: Count documents in collection
    println!("\n=== Count Documents ===");
    let count = client.count_documents(collection).await?;
    println!("Document count: {}", count);

    // Example 4: Check if collection exists
    println!("\n=== Check Collection Exists ===");
    let exists = client.collection_exists(collection).await?;
    println!("Collection exists: {}", exists);

    // Example 5: Delete collection
    println!("\n=== Delete Collection ===");
    client.delete_collection(collection).await?;
    println!("Collection deleted successfully");

    // Example 6: Verify deletion
    println!("\n=== Verify Deletion ===");
    let still_exists = client.collection_exists(collection).await?;
    println!("Collection still exists: {}", still_exists);

    println!("\n✓ All collection management operations completed successfully");

    Ok(())
}

File: examples/kotlin/examples/ClientCollectionManagement.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Collection management example - Create, list, and manage collections
 */
fun main() = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"
    
    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()
    
    val collection = "kotlin_collection_example"
    
    println("=== ekoDB Kotlin Client - Collection Management Example ===\n")
    
    try {
        // Example 1: List all collections
        println("=== List Collections ===")
        val collections = client.listCollections()
        println("✓ Found ${collections.size} collections")
        collections.take(5).forEach { println("  - $it") }
        println()
        
        // Example 2: Check if collection exists
        println("=== Check Collection Existence ===")
        val exists = client.collectionExists(collection)
        println("Collection '$collection' exists: $exists\n")
        
        // Example 3: Get collection metadata (if exists)
        if (exists) {
            println("=== Get Collection Metadata ===")
            try {
                val metadata = client.getCollection(collection)
                println("✓ Collection metadata:")
                println("  $metadata\n")
            } catch (e: Exception) {
                println("⚠ Could not get metadata: ${e.message}\n")
            }
        }
        
        // Example 4: Delete collection if exists
        if (exists) {
            println("=== Delete Collection ===")
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection\n")
        }
        
        // Example 5: Create collection with schema
        println("=== Create Collection with Schema ===")
        val schema = buildJsonObject {
            put("fields", buildJsonObject {
                put("name", buildJsonObject {
                    put("field_type", "String")
                    put("required", true)
                })
                put("age", buildJsonObject {
                    put("field_type", "Integer")
                    put("required", false)
                })
            })
        }
        
        client.createCollection(collection, schema)
        println("✓ Created collection with schema: $collection\n")
        
        // Example 6: Get schema
        println("=== Get Collection Schema ===")
        val retrievedSchema = client.getSchema(collection)
        println("✓ Schema: $retrievedSchema\n")
        
    } finally {
        // Cleanup
        println("=== Cleanup ===")
        try {
            client.deleteCollection(collection)
            println("✓ Deleted collection: $collection")
        } catch (e: Exception) {
            println("⚠ Could not delete collection: ${e.message}")
        }
        
        client.close()
        println("\n=== Example Complete ===")
    }
}
